design/one-pager-k8s-native-providers.md

```diff
@@ -0,0 +1,271 @@
+operator into a cluster, you should be able to dynamically provision a Minio
+object storage cluster into the Kubernetes cluster from your Crossplane control
+cluster. The experience would look as follows:
+
+1. Create a non-portable Rook Minio `ObjectStoreClass`
+
+```yaml
+apiVersion: minio.rook.crossplane.io/v1alpha1
+kind: ObjectStoreClass
+metadata:
+  name: rook-minio
+  namespace: rook-infra-dev
+spec:
+  targetClusterRef:
+    apiVersion: workload.crossplane.io/v1alpha1
+    kind: KubernetesCluster
+    name: my-remote-cluster
+    namespace: rook-infra-dev
+  scope:
+    nodeCount: 4
+    volumeClaimTemplates:
+    - metadata:
+        name: rook-minio-data1
+      spec:
+        accessModes: [ "ReadWriteOnce" ]
+        resources:
+          requests:
+            storage: "8Gi"
+  # A key value list of annotations
+  annotations:
+  #  key: value
+  placement:
+    tolerations:
+    nodeAffinity:
+    podAffinity:
+    podAnyAffinity:
+  credentials:
+    name: minio-my-store-access-keys
+    namespace: rook-minio
+```
+
+2. Create a portable `BucketClass`
+
+```yaml
+apiVersion: storage.crossplane.io/v1alpha1
+kind: BucketClass
+metadata:
+  name: bucket-standard
+  namespace: app-project1-dev
+classRef:
+  kind: ObjectStoreClass
+  apiVersion: minio.rook.crossplane.io/v1alpha1
+  name: rook-minio
+  namespace: rook-infra-dev
+```
+
+3. Create a `Bucket` claim
+
+```yaml
+apiVersion: storage.crossplane.io/v1alpha1
+kind: Bucket
+metadata:
+  name: app-bucket
+  namespace: app-project1-dev
+spec:
+  classRef:
+    name: bucket-standard
+  writeConnectionSecretToRef:
+    name: bucketsecret
+```
+
+On creation of the `Bucket` claim, the claim controller would configure and
+create an `ObjectStore` (`minio.rook.crossplane.io/v1alpha1`) instance. Then,
+the managed controller would use the Crossplane `ObjectStore` instance to create
+an `ObjectStore` (`minio.rook.io/v1alpha1`) in the target Kubernetes cluster
+where Rook and the Minio operator are installed. This workflow is similar to
+dynamic provisioning in a typical infrastructure provider:
+
+![K8s Native Providers](./images/k8s-native-providers.png)
+
+## Technical Design
+
+### Claim Reconciler
+
+The claim controllers could use the same shared [claim reconciler] that other
+existing infrastructure stacks use currently, providing their own
+`ManagedConfigurators` to configure the managed resource instance using the
+referenced class.
+
+### Managed Reconciler
+
+Kubernetes-native infrastructure providers can make use of the shared managed
+reconciler in `crossplane-runtime` in the same manner as traditional
+infrastructure providers. However, instead of creating a new client for a cloud
+provider API, Kubernetes-native infrastructure providers will return a
+Kubernetes client that is configured to talk to the target cluster using the
+`targetClusterRef`. This is similar to the client configuration for the
+`KubernetesApplicationResource` [managed reconciler].
```

Comment:
Does this mean these providers use a `KubernetesCluster` claim rather than a `Provider`? How will we solve the cases mentioned in https://github.com/crossplaneio/crossplane/pull/841#pullrequestreview-293432522?

Reply:
Yes, the current implementation here suggests using a `KubernetesCluster` ref and deferring that effort to #497. However, implementing a new `Provider` would fix these issues for Rook and be more in-line with how existing infrastructure stacks operate. Do you think that is the appropriate direction here?